src/Npgsql/NpgsqlRecordDataReader.cs

```diff
@@ -0,0 +1,327 @@
+﻿using Npgsql.Internal;
+using Npgsql.Internal.TypeHandling;
+using Npgsql.PostgresTypes;
+using Npgsql.TypeMapping;
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Data.Common;
+using System.Diagnostics;
+using System.IO;
+using System.Runtime.CompilerServices;
+using System.Text;
+
+namespace Npgsql
+{
+    internal sealed class NpgsqlRecordDataReader : DbDataReader
```

Comment:
Maybe NpgsqlInnerDataReader? For now it can handle both a single ROW() and an array of `ROW()`s.
Note that the name shouldn't matter that much since, for now, the class is internal.

Reply:
Yeah, agree it doesn't matter much - NpgsqlInnerDataReader sounds good.

Reply:
Ended up going with NpgsqlNestedDataReader